hw9/hw9_t01.py

```diff
@@ -17,32 +17,51 @@
 from typing import Iterator, List, Union
 
 
-def open_file_return_list(path: Union[Path, str]) -> list:
-    with open(path, "r") as file:
-        return list(map(int, file.read().split()))
+def open_file_return_iterator(path: Union[Path, str]) -> Iterator:
+    with open(path) as file:
+        yield from (int(row) for row in file)
 
 
-def merge_2_sorted_lists(first: list, second: list) -> list:
-    i = k = 0
+def merge_2_sorted_iterators(  # noqa: C901,CCR001
+    left: Iterator, right: Iterator
+) -> Iterator:
     result = []
-    while i < len(first) and k < len(second):
-        if first[i] <= second[k]:
-            result.append(first[i])
-            i += 1
-        else:
-            result.append(second[k])
-            k += 1
+    try:
+        left_item = next(left)
+    except StopIteration:
+        result.extend(right)
+        return iter(result)
 
-    result.extend(first[i:])
-    result.extend(second[k:])
+    try:
+        right_item = next(right)
+    except StopIteration:
+        result.extend(left)
+        return iter(result)
 
-    return result
+    while True:
+        if left_item <= right_item:
+            result.append(left_item)
+            try:
+                left_item = next(left)
+            except StopIteration:
+                result.append(right_item)
+                result.extend(right)
+                break
```

Comment:
Tried to take it out to a dedicated function: code got less clear in my opinion

Reply:
You can move both extend to the end.  This will simplify code a little.  

You can try to play with the next arguments. 
You can invent your own data structure that will be easy to use and it will make code simplier.